code/parachain/frame/pablo/src/testing.rs

```diff
@@ -0,0 +1,297 @@
+#![allow(clippy::unwrap_used, clippy::disallowed_methods)]
+
+use std::collections::BTreeMap;
+
+use composable_support::math::safe::SafeAdd;
+use composable_tests_helpers::{
+	test::{
+		block::next_block,
+		currency::{Currency, PICA, USDT},
+		helper::RuntimeTrait,
+	},
+	ALICE, BOB,
+};
+use composable_traits::currency::BalanceLike;
+use frame_support::{
+	assert_noop, bounded_btree_map,
+	traits::{fungibles::Mutate, OriginTrait, TryCollect},
+};
+use frame_system::{pallet_prelude::OriginFor, Config as SystemConfig};
+use sp_arithmetic::Permill;
+use sp_runtime::{traits::Zero, AccountId32};
+
+use crate::{Config, Event, Pallet, PoolInitConfiguration};
+
+#[allow(clippy::upper_case_acronyms)]
+type KSM = Currency<4, 12>;
+
+// NOTE/FIXME(benluelo): These trait bounds can be simplified quite a bit once this issue is
+// resolved: https://github.com/rust-lang/rust/issues/20671#issuecomment-529752828
+//
+// Associated type constraints currently aren't resolved at the use site, but super traits (and
+// their associated type constraints) are! To work with this, we write some extra boilerplate here
+// to make the call site only require a simple `<Runtime: Trait>` constraint as opposed to repeating
+// the trait bounds on every function.
+
+/// This trait defines the constraints that pablo assumes from the runtime, allowing tests using
+/// this trait to bound the runtime to be run with any runtime thatr fits within these criteria.
+pub trait PabloRuntimeConstraints:
+	RuntimeTrait<Event<Self>>
+	+ SystemConfig<BlockNumber = Self::__SystemBlockNumber, AccountId = Self::__SystemAccountId>
+	+ Config<Balance = Self::__PabloBalance> // , AssetId = Self::__AssetId
+	+ pallet_timestamp::Config<Moment = Self::__TimestampMoment>
+{
+	// these associated types aren't intended to be used, they only exist for the reasons
+	// described above.
+	type __SystemBlockNumber: From<u32> + Into<u64> + SafeAdd;
+	type __SystemAccountId: From<AccountId32> + Clone;
+	type __TimestampMoment: From<u64>;
+	type __PabloBalance: BalanceLike + From<u128> + Zero;
+	// type __AssetId: Clone;
+}
+
+impl<T> PabloRuntimeConstraints for T
+where
+	T: RuntimeTrait<Event<Self>>,
```

Comment:
hm. seems it may work for similar case in production code in cosmwasm pallet

Reply:
it's a nifty trick, can save a lot of boilerplate if you use the same constraints a lot. there are some limitations with it though that i'm documenting as well

Reply:
Why not just flip the bound around?

```
where Runtime: pablo::Config?
```